docs/hugo/content/contributing/add-a-new-code-generated-resource/write-a-test.md

```diff
@@ -0,0 +1,49 @@
+weight: 40
+---
+
+Every new resource should have a handwritten test as there is always the possibility of a resource behaving in a subtly different way than expected. This also applies to _new versions_ of existing resources.
+
+All these tests live in the [`v2/internal/controllers`](https://github.com/Azure/azure-service-operator/tree/main/v2/internal/controllers). There's a readme in that folder that describes the test structure and naming conventions.
+
+For each test there's a recording in the `recordings` folder that records the HTTP interactions between the test and Azure allowing for rapid replay for future runs.
+
+One good approach is to start from an [existing test](https://github.com/Azure/azure-service-operator/blob/main/v2/internal/controllers/documentdb_mongodb_crud_v20231115_test.go) and modify it to work for your resource. It can also be helpful to refer to examples in the [ARM templates GitHub repo](https://github.com/Azure/azure-quickstart-templates).
+
+## Resource extensions
+
+There are some variations in the way different product groups implement their resources. To support these variations, there are a number of available extension points for resources that allow the behaviour to be modified. 
+
+For example, some resources will return a `BadRequest` error when one prerequisite is not met. Normally, ASO will treat `BadRequest` as a permanent failure, but this can be reclassified as a warning by using the [`ErrorClassifier`](https://github.com/Azure/azure-service-operator/blob/main/v2/pkg/genruntime/extensions/error_classifier.go#L19) extension point.
+
+Other extension points allow blocking reconciliation attempts until a prerequisite is met, additional API calls after a successful reconciliation, or even modifying the resource before it is sent to Azure.
+
+For most resource extensions, any implementation will go in the `customizations` folder generated for each resource group.
+
+Reach out to us if you want some guidance on which (if any) extension points are appropriate for your resource.
+
+## Record the test passing
+
+When the test passes, you'll find a new recording file in the `recordings` folder. When you modify your test, you'll need to delete the recording file so that the test runs against the real Azure API. 
+
+Once you're finished, include the recording file in your PR. This is important because it allows other developers to run the test without needing to make real API calls to Azure.
+
+See [the code generator test README](../testing/#running-envtest-integration-tests) for more information on how to run tests and record their HTTP interactions to allow replay.
+
+## Test volumes
+
+These integration tests can take a few minutes to run from recordings, and we're starting to see issues due to the large number of resources we're testing. 
+
+If you are adding a new version of an already supported resource, we may request that you remove tests for older versions of the resource as a part of your PR. 
```

Comment:
Is that a good idea if the previous versions are still in use and someone in a service team mucks something up and introduces a breaking change?

Reply:
We're reluctant to remove older tests for exactly this reason, but our CI times are starting to climb unreasonably so we need to do something.

Reply:
FWIW the replay tests when run in replay mode aren't the best for protecting against service teams making breaking changes anyway because we'll always replay the old (non-broken) payloads.

Removing old tests does suck too, but 1 test per API per resource for all of Azure is a lot of tests. As we support more and more resources and versions something has to give in the matrix.